{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.feature_extraction.text import CountVectorizer\n",
    "from sklearn.feature_extraction.text import TfidfTransformer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                        Title  \\\n",
      "Project ID                                                      \n",
      "SRP099026                sludge metagenome Raw sequence reads   \n",
      "SRP107999   Aqueous microbial communities from the Delawar...   \n",
      "SRP080517   Grasslands soil microbial communities from the...   \n",
      "SRP007817   Hoatzin crop microbiome epithelium fraction 12...   \n",
      "ERP021394   Environmental alkali multi-extreme Diamante La...   \n",
      "SRP096603   Active sludge microbial communities of municip...   \n",
      "ERP006694                                      test sequences   \n",
      "SRP100041   Marine microbial communities from the Southern...   \n",
      "SRP006444   Community genomic analysis of an extremely aci...   \n",
      "SRP104952   Marine viral communities from the Global Malas...   \n",
      "SRP041163                  Bovine rumen microbiome Metagenome   \n",
      "SRP099977   Freshwater microbial communities from Crystal ...   \n",
      "SRP059473     Dechlorination Culture CG-3 and SG-1 Metagenome   \n",
      "SRP081889   Freshwater microbial communities from Trout Bo...   \n",
      "SRP091673                          17th century variola virus   \n",
      "SRP104826   Estuarine microbial communities from the Colum...   \n",
      "SRP041649                        Acidic hot spring Metagenome   \n",
      "SRP098105   Estuarine microbial communities from the Colum...   \n",
      "SRP081953   Freshwater microbial communities from Trout Bo...   \n",
      "SRP003198                     Metagenome isolated from kimchi   \n",
      "SRP081962   Freshwater microbial communities from Trout Bo...   \n",
      "SRP104928   Freshwater sediment microbial communities from...   \n",
      "SRP066089   Deep metagenome and transcriptome analyses of ...   \n",
      "SRP099516   Avena fatua rhizosphere microbial communities ...   \n",
      "SRP098069   Groundwater microbial communities from the Col...   \n",
      "SRP081480   Wastewater treatment Type I Accumulibacter com...   \n",
      "SRP092476   Synthetic human gut microbiota 14 species asse...   \n",
      "SRP100226   Bog forest soil microbial communities from Cal...   \n",
      "SRP077422   Marine microbial communities from expanding ox...   \n",
      "SRP100314   Freshwater lake sediment microbial communities...   \n",
      "...                                                       ...   \n",
      "SRP103438   Freshwater sediment methanotrophic microbial c...   \n",
      "SRP100025   Freshwater microbial communities from Crystal ...   \n",
      "SRP046072   Pit mud of Chinese liqour fermentation reactor...   \n",
      "SRP080037   Marine microbial communities from the Deep Pac...   \n",
      "SRP080066   Marine microbial communities from the Deep Atl...   \n",
      "SRP104998   Groundwater sediment microbial communities fro...   \n",
      "SRP080527   Arabidopsis root microbial communities from th...   \n",
      "SRP103417   Freshwater sediment methanotrophic microbial c...   \n",
      "SRP090270                          soil metagenome Metagenome   \n",
      "ERP022764   A case of hepatic brucelloma studied by next g...   \n",
      "SRP116264                    Human Milk Microbiome Metagenome   \n",
      "SRP099916   Freshwater lake microbial communities from Lak...   \n",
      "SRP100267   Soil microbial communities from Arlington Agri...   \n",
      "SRP024366   RNA and DNA based 16S rRNA analysis of marine ...   \n",
      "ERP021528   study of microbial communities of a lake durin...   \n",
      "SRP043136                Bacterial communities of ctenophores   \n",
      "SRP100116   Marine microbial communities from expanding ox...   \n",
      "SRP106062   Marine microbial communities from the Southern...   \n",
      "ERP010513   Chinese Poyang Lake resistant genes from soil ...   \n",
      "SRP061540   Biodiversity and functioning of microbial comm...   \n",
      "ERP016524            International Human Microbiome Standards   \n",
      "SRP040735   Are abundant bacteria more active than rare ba...   \n",
      "SRP080278   Freshwater microbial communities from Lake Men...   \n",
      "ERP008846   Eukaryotic mRNA from 1 year old litter in beec...   \n",
      "SRP098511   Saline lake microbial communities from Deep la...   \n",
      "SRP045196                               Wastewater metagenome   \n",
      "SRP010824   Targeted Biofilm samples from two redox zones-...   \n",
      "SRP080431   Soil microbial communities from Arlington Agri...   \n",
      "SRP099951   Freshwater microbial communities from Crystal ...   \n",
      "SRP080508   Freshwater and sediment microbial communities ...   \n",
      "\n",
      "                                                     Abstract  \\\n",
      "Project ID                                                      \n",
      "SRP099026                Isolation of phages from waste water   \n",
      "SRP107999   Biogeochemical cycling links between terrestri...   \n",
      "SRP080517                                                None   \n",
      "SRP007817                                                None   \n",
      "ERP021394   Diamante Lake is a multi-extreme environment l...   \n",
      "SRP096603   Metagenomics of anaerobic wastewater microbial...   \n",
      "ERP006694                                        workshop aau   \n",
      "SRP100041                                                None   \n",
      "SRP006444   Metagenomic analysis of highly acidic pH 0-1 b...   \n",
      "SRP104952   Identifying viral communities during a global ...   \n",
      "SRP041163   In this study we sought to explore the fibroly...   \n",
      "SRP099977                                                None   \n",
      "SRP059473                  Metagenomic data for CG-3 and SG-1   \n",
      "SRP081889                                                None   \n",
      "SRP091673   Recovery of variola virus through targeted enr...   \n",
      "SRP104826   Sequencing of microbial communities from the C...   \n",
      "SRP041649   Metagenomic raw reads from acidic hot spring m...   \n",
      "SRP098105   Sequencing of microbial communities from the C...   \n",
      "SRP081953                                                None   \n",
      "SRP003198   Microbiomes isolated from Korean fermented foo...   \n",
      "SRP081962                                                None   \n",
      "SRP104928   Metagenome sequencing of total DNA extracted f...   \n",
      "SRP066089   A study to investigate differences in the hydr...   \n",
      "SRP099516                                                None   \n",
      "SRP098069   Coupling Microbial Communities to Carbon and C...   \n",
      "SRP081480                                                None   \n",
      "SRP092476   The goal of this study was to determine transc...   \n",
      "SRP100226                                                None   \n",
      "SRP077422                                                None   \n",
      "SRP100314                                                None   \n",
      "...                                                       ...   \n",
      "SRP103438   Methane oxidation as a community function: Fre...   \n",
      "SRP100025                                                None   \n",
      "SRP046072   Chinese liquor is produced by a solid-state fe...   \n",
      "SRP080037                                                None   \n",
      "SRP080066                                                None   \n",
      "SRP104998   Genome-resolved metagenomic analysis of microb...   \n",
      "SRP080527                                                None   \n",
      "SRP103417   Methane oxidation as a community function: Fre...   \n",
      "SRP090270   Microorganisms have an essential role in earth...   \n",
      "ERP022764   A clinical examination of a patient pointed to...   \n",
      "SRP116264   Maternal diet during lactation is known to inf...   \n",
      "SRP099916                                                None   \n",
      "SRP100267                                                None   \n",
      "SRP024366   Marine bacterioplankton samples were collected...   \n",
      "ERP021528   three different depth points sediment and wate...   \n",
      "SRP043136   Based on Automated Ribosomal Intergenic Spacer...   \n",
      "SRP100116                                                None   \n",
      "SRP106062   Microbial engines driving organic matter trans...   \n",
      "ERP010513   Chinese Poyang Lake resistant genes from soil ...   \n",
      "SRP061540   Microorganisms and microbial associations play...   \n",
      "ERP016524   The IHMS EU funded project aimed at coordinate...   \n",
      "SRP040735   The project will use a combination of pyrosequ...   \n",
      "SRP080278                                                None   \n",
      "ERP008846   Metatranscriptommic study of eukyryotic microb...   \n",
      "SRP098511   Metagenomic analysis of microbial communities ...   \n",
      "SRP045196   Wastewater contains a diverse array of toxic c...   \n",
      "SRP010824                                                None   \n",
      "SRP080431                                                None   \n",
      "SRP099951                                                None   \n",
      "SRP080508                                                None   \n",
      "\n",
      "                                                         Runs  \n",
      "Project ID                                                     \n",
      "SRP099026                                          SRR5236411  \n",
      "SRP107999                                          SRR5601898  \n",
      "SRP080517                                SRR3989441SRR3989442  \n",
      "SRP007817   SRR351327SRR351328SRR446578SRR446579SRR446580S...  \n",
      "ERP021394                                          ERR1824222  \n",
      "SRP096603                                          SRR5168925  \n",
      "ERP006694                                           ERR580960  \n",
      "SRP100041                                          SRR5262284  \n",
      "SRP006444                                           SRR189081  \n",
      "SRP104952                                          SRR5468455  \n",
      "SRP041163                                          SRR1239316  \n",
      "SRP099977                                          SRR5260837  \n",
      "SRP059473                      SRR2062390SRR2062392SRR2062395  \n",
      "SRP081889                                          SRR4029080  \n",
      "SRP091673                                          SRR4427811  \n",
      "SRP104826                                          SRR5468158  \n",
      "SRP041649                                SRR1297186SRR1297204  \n",
      "SRP098105                                          SRR5209940  \n",
      "SRP081953                                          SRR4029393  \n",
      "SRP003198       SRR256747SRR256748SRR256749SRR256751SRR256756  \n",
      "SRP081962                                          SRR4029402  \n",
      "SRP104928                                          SRR5468427  \n",
      "SRP066089            SRR2917896SRR2917897SRR2917898SRR2917899  \n",
      "SRP099516                                          SRR5252352  \n",
      "SRP098069                                          SRR5209622  \n",
      "SRP081480                                          SRR4023301  \n",
      "SRP092476                                SRR4841891SRR4841892  \n",
      "SRP100226                                          SRR5271166  \n",
      "SRP077422                                          SRR3724176  \n",
      "SRP100314                                          SRR5272498  \n",
      "...                                                       ...  \n",
      "SRP103438                                          SRR5431137  \n",
      "SRP100025                                          SRR5261471  \n",
      "SRP046072   SRR1561950SRR1561989SRR1561990SRR1561991SRR156...  \n",
      "SRP080037                                          SRR3962773  \n",
      "SRP080066                                          SRR3965874  \n",
      "SRP104998                                          SRR5468955  \n",
      "SRP080527                                SRR3989452SRR3989453  \n",
      "SRP103417                                          SRR5430755  \n",
      "SRP090270   SRR4289280SRR4289282SRR4289283SRR4289284SRR428...  \n",
      "ERP022764            ERR1951438ERR1951439ERR1951440ERR1951441  \n",
      "SRP116264   SRR5980292SRR5980293SRR5980294SRR5980295SRR598...  \n",
      "SRP099916                      SRR5260446SRR5260447SRR5260448  \n",
      "SRP100267                                          SRR5272127  \n",
      "SRP024366   SRR3425686SRR3425688SRR3425693SRR3425695SRR342...  \n",
      "ERP021528                      ERR1843312ERR1843313ERR1843314  \n",
      "SRP043136            SRR1391031SRR1391032SRR1391060SRR1391071  \n",
      "SRP100116                                          SRR5265135  \n",
      "SRP106062                                          SRR5495982  \n",
      "ERP010513   ERR970587ERR970588ERR970589ERR970590ERR970591E...  \n",
      "SRP061540            SRR2125895SRR2163617SRR5626535SRR5626536  \n",
      "ERP016524   ERR1539337ERR1539338ERR1539339ERR1539340ERR153...  \n",
      "SRP040735   SRR1211819SRR1211842SRR1211874SRR1211893SRR123...  \n",
      "SRP080278                                          SRR3985708  \n",
      "ERP008846   ERR688960ERR688963ERR688966ERR688968ERR688971E...  \n",
      "SRP098511                                          SRR5216651  \n",
      "SRP045196   SRR1535019SRR1535020SRR1535021SRR1535022SRR153...  \n",
      "SRP010824                                  SRR407181SRR407182  \n",
      "SRP080431                                          SRR3988847  \n",
      "SRP099951                                          SRR5260685  \n",
      "SRP080508                                          SRR3989431  \n",
      "\n",
      "[5385 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv('WGS_abstracts.txt', sep = '\\t', names = ['Project ID','Title','Abstract','Runs'])\n",
    "df.set_index('Project ID', inplace=True)\n",
    "\n",
    "df = df.replace(r'^\\s+$', 'None', regex=True)\n",
    "df = df.replace(r'none provided', 'None', regex=True)\n",
    "df = df.replace(r'â€˜', '', regex=True)\n",
    "df = df.replace(r'â€™', '', regex=True)\n",
    "df = df.replace(r'\\.(?!\\d)', '', regex=True)\n",
    "df = df.replace(r',', '', regex=True)\n",
    "df = df.replace(r'\\(', '', regex=True)\n",
    "df = df.replace(r'\\)', '', regex=True)\n",
    "df = df.replace(np.nan, '')\n",
    "print(df)\n",
    "\n",
    "abstracts = df['Abstract']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "cvec = CountVectorizer(stop_words='english', min_df=1, max_df=.25, ngram_range=(1,2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('isolation', 993),\n",
       " ('phages', 1355),\n",
       " ('water', 1967),\n",
       " ('biogeochemical', 239),\n",
       " ('cycling', 474),\n",
       " ('links', 1064),\n",
       " ('terrestrial', 1846),\n",
       " ('marine', 1099),\n",
       " ('systems', 1819),\n",
       " ('biogeochemical cycling', 241),\n",
       " ('cycling links', 477),\n",
       " ('links terrestrial', 1065),\n",
       " ('terrestrial marine', 1847),\n",
       " ('marine systems', 1104),\n",
       " ('lake', 1024),\n",
       " ('multi', 1231),\n",
       " ('extreme', 703),\n",
       " ('environment', 649),\n",
       " ('located', 1073),\n",
       " ('developing', 535)]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from itertools import islice\n",
    "cvec.fit(abstracts)\n",
    "list(islice(cvec.vocabulary_.items(), 20))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5385, 2000)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cvec_counts = cvec.transform(abstracts)\n",
    "cvec_counts.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>term</th>\n",
       "      <th>occurrences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>communities</td>\n",
       "      <td>1182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1650</th>\n",
       "      <td>sequencing</td>\n",
       "      <td>1085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>community</td>\n",
       "      <td>891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1762</th>\n",
       "      <td>study</td>\n",
       "      <td>876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1136</th>\n",
       "      <td>metagenomic</td>\n",
       "      <td>848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1180</th>\n",
       "      <td>microbial communities</td>\n",
       "      <td>809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1605</th>\n",
       "      <td>samples</td>\n",
       "      <td>770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1702</th>\n",
       "      <td>soil</td>\n",
       "      <td>672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284</th>\n",
       "      <td>carbon</td>\n",
       "      <td>613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>569</th>\n",
       "      <td>dna</td>\n",
       "      <td>609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1596</th>\n",
       "      <td>sample</td>\n",
       "      <td>594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1131</th>\n",
       "      <td>metagenome</td>\n",
       "      <td>592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>analysis</td>\n",
       "      <td>539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1929</th>\n",
       "      <td>using</td>\n",
       "      <td>522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>854</th>\n",
       "      <td>gut</td>\n",
       "      <td>473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>bacterial</td>\n",
       "      <td>469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565</th>\n",
       "      <td>diversity</td>\n",
       "      <td>457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1199</th>\n",
       "      <td>microbiota</td>\n",
       "      <td>437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1195</th>\n",
       "      <td>microbiome</td>\n",
       "      <td>434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>343</th>\n",
       "      <td>collected</td>\n",
       "      <td>382</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       term  occurrences\n",
       "359             communities         1182\n",
       "1650             sequencing         1085\n",
       "374               community          891\n",
       "1762                  study          876\n",
       "1136            metagenomic          848\n",
       "1180  microbial communities          809\n",
       "1605                samples          770\n",
       "1702                   soil          672\n",
       "284                  carbon          613\n",
       "569                     dna          609\n",
       "1596                 sample          594\n",
       "1131             metagenome          592\n",
       "128                analysis          539\n",
       "1929                  using          522\n",
       "854                     gut          473\n",
       "208               bacterial          469\n",
       "565               diversity          457\n",
       "1199             microbiota          437\n",
       "1195             microbiome          434\n",
       "343               collected          382"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ = np.asarray(cvec_counts.sum(axis=0)).ravel().tolist()\n",
    "counts_df = pd.DataFrame({'term': cvec.get_feature_names(), 'occurrences': occ})\n",
    "counts_df.sort_values(by='occurrences', ascending=False).head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "transformer = TfidfTransformer()\n",
    "transformed_weights = transformer.fit_transform(cvec_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>term</th>\n",
       "      <th>weight</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>communities</td>\n",
       "      <td>0.024028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1650</th>\n",
       "      <td>sequencing</td>\n",
       "      <td>0.021107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1180</th>\n",
       "      <td>microbial communities</td>\n",
       "      <td>0.019875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1136</th>\n",
       "      <td>metagenomic</td>\n",
       "      <td>0.018313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1762</th>\n",
       "      <td>study</td>\n",
       "      <td>0.017185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>community</td>\n",
       "      <td>0.017013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1702</th>\n",
       "      <td>soil</td>\n",
       "      <td>0.016432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1131</th>\n",
       "      <td>metagenome</td>\n",
       "      <td>0.016332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1605</th>\n",
       "      <td>samples</td>\n",
       "      <td>0.013851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>analysis</td>\n",
       "      <td>0.013447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284</th>\n",
       "      <td>carbon</td>\n",
       "      <td>0.013274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1596</th>\n",
       "      <td>sample</td>\n",
       "      <td>0.011570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1195</th>\n",
       "      <td>microbiome</td>\n",
       "      <td>0.011022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1148</th>\n",
       "      <td>metagenomics</td>\n",
       "      <td>0.010296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>474</th>\n",
       "      <td>cycling</td>\n",
       "      <td>0.009889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>569</th>\n",
       "      <td>dna</td>\n",
       "      <td>0.009829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>bacterial</td>\n",
       "      <td>0.009608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>854</th>\n",
       "      <td>gut</td>\n",
       "      <td>0.009247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565</th>\n",
       "      <td>diversity</td>\n",
       "      <td>0.009211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1628</th>\n",
       "      <td>sediment</td>\n",
       "      <td>0.009073</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       term    weight\n",
       "359             communities  0.024028\n",
       "1650             sequencing  0.021107\n",
       "1180  microbial communities  0.019875\n",
       "1136            metagenomic  0.018313\n",
       "1762                  study  0.017185\n",
       "374               community  0.017013\n",
       "1702                   soil  0.016432\n",
       "1131             metagenome  0.016332\n",
       "1605                samples  0.013851\n",
       "128                analysis  0.013447\n",
       "284                  carbon  0.013274\n",
       "1596                 sample  0.011570\n",
       "1195             microbiome  0.011022\n",
       "1148           metagenomics  0.010296\n",
       "474                 cycling  0.009889\n",
       "569                     dna  0.009829\n",
       "208               bacterial  0.009608\n",
       "854                     gut  0.009247\n",
       "565               diversity  0.009211\n",
       "1628               sediment  0.009073"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "weights = np.asarray(transformed_weights.mean(axis=0)).ravel().tolist()\n",
    "weights_df = pd.DataFrame({'term': cvec.get_feature_names(), 'weight': weights})\n",
    "weights_df.sort_values(by='weight', ascending=False).head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_train = transformed_weights\n",
    "mu = x_train.mean(axis=0)\n",
    "U,s,V = np.linalg.svd(x_train - mu, full_matrices=False)\n",
    "Zpca = np.dot(x_train - mu, V.transpose())\n",
    "\n",
    "Rpca = np.dot(Zpca[:,:2], V[:2,:]) + mu    # reconstruction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "err = np.sum(np.asarray(x_train-Rpca)**2)/Rpca.shape[0]/Rpca.shape[1]\n",
    "print(err)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter([Zpca[:2000,0]], [Zpca[:2000,1]])\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
